component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+			return 0, err
+		}
+	}
+	if rw.StreamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().ReadInBuffer(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	if atomic.LoadInt64(&options.Handle.CacheObj.Size) == 0 {
+		return 0, nil
+	}
+	read, err := rw.readWriteBlocks(options.Handle, options.Offset, options.Data, false)
+	if err != nil {
+		log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
+	}
+	return read, err
+}
+
+func (rw *ReadWriteFilenameCache) WriteFile(options internal.WriteFileOptions) (int, error) {
+	// log.Trace("Stream::WriteFile : name=%s, handle=%d, offset=%d", options.Handle.Path, options.Handle.ID, options.Offset)
+	if !rw.StreamOnly && options.Handle.CacheObj.StreamOnly {
+		err := rw.createFileCache(options.Handle)
+		if err != nil {
+			log.Err("Stream::WriteFile : error failed to create cache object %s [%s]", options.Handle.Path, err.Error())
+			return 0, err
+		}
+	}
+	if rw.StreamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().WriteFile(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::WriteFile : error failed to write data to %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	written, err := rw.readWriteBlocks(options.Handle, options.Offset, options.Data, true)
+	if err != nil {
+		log.Err("Stream::WriteFile : error failed to write data to %s: [%s]", options.Handle.Path, err.Error())
+	}
+	options.Handle.Flags.Set(handlemap.HandleFlagDirty)
+	return written, err
+}
+
+// TODO: truncate in cache
+func (rw *ReadWriteFilenameCache) TruncateFile(options internal.TruncateFileOptions) error {
+	// log.Trace("Stream::TruncateFile : name=%s, size=%d", options.Name, options.Size)
+	if !rw.StreamOnly {
+		rw.purge(options.Name, false)
+	}
+	err := rw.NextComponent().TruncateFile(options)
+	if err != nil {
+		log.Err("Stream::TruncateFile : error truncating file %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) RenameFile(options internal.RenameFileOptions) error {
+	// log.Trace("Stream::RenameFile : name=%s", options.Src)
+	if !rw.StreamOnly {
+		rw.purge(options.Src, false)
+	}
+	err := rw.NextComponent().RenameFile(options)
+	if err != nil {
+		log.Err("Stream::RenameFile : error renaming file %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (rw *ReadWriteFilenameCache) CloseFile(options internal.CloseFileOptions) error {
+	// log.Trace("Stream::CloseFile : name=%s, handle=%d", options.Handle.Path, options.Handle.ID)
+	// try to flush again to make sure it's cleaned up
+	err := rw.FlushFile(internal.FlushFileOptions(options))
```

Comment:
Only when handle is dirty we shall call the flush file

Reply:
I'm leaving it up to storage component to make that check, do you think we should just do it here?

Reply:
storage component we do not make any dirty check (atleast for file-cache). As handle or block is marked dirty by streaming ideally respective checks shall be done by the layer itself and not upstream component. Storage shall be dumb component, if someone asks it to upload it shall just do it.

Reply:
Done